File: app/src/main/java/com/dsy/mvp/base/BaseActivity.java
package com.dsy.mvp.base;

import android.app.Activity;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.blankj.utilcode.util.ToastUtils;
import com.dsy.mvp.base.impl.IPresenter;
import com.dsy.mvp.base.impl.IView;
import com.dsy.mvp.component.ActivityManager;

import butterknife.ButterKnife;

public abstract class BaseActivity<T extends IPresenter> extends AppCompatActivity implements IView {

    protected final String TAG = this.getClass().getSimpleName();
    public ProgressDialog progressDialog;
    protected T mPresenter;
    protected Bundle savedInstanceState;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.savedInstanceState = savedInstanceState;
        beforeCreate();
        if (getLayoutId() != 0) {
            setContentView(getLayoutId());
        }
        ActivityManager.getInstance().addActivity(this);
        ButterKnife.bind(this);
        mPresenter = createPresenter();
        attachView();
        initView();
        initEvent();
        initData();
    }

    /**
     * 布局载入之前
     */
    protected void beforeCreate() {
    }

    /**
     * 布局载入
     */
    protected abstract int getLayoutId();

    /**
     * P层绑定   若无则返回null;
     */
    protected abstract T createPresenter();

    /**
     * 控件绑定
     */
    protected abstract void initView();

    /**
     * 事件触发绑定
     */
    protected abstract void initEvent();

    /**
     * 数据初始化
     */
    protected abstract void initData();

    protected void setToolBar(Toolbar toolbar, String title) {
        toolbar.setTitle(title);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });
    }

    /**
     * P层绑定V层
     */
    private void attachView() {
        if (null != mPresenter) {
            mPresenter.attachView(this);
        }
    }

    /**
     * P层解绑V层
     */
    private void detachView() {
        if (null != mPresenter) {
            mPresenter.detachView();
        }
    }

    @Override
    public Activity getMActivity() {
        return this;
    }

    @Override
    public void toast(int id) {
        ToastUtils.showShort(id);
    }

    @Override
    public void toast(String msg) {
        ToastUtils.showShort(msg);
    }

    @Override
    public void showLoadingDialog() {
        progressDialog = new ProgressDialog(this);
        progressDialog.setMessage("加载中");
        progressDialog.show();
    }

    @Override
    public void hideLoadingDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        detachView();
        ActivityManager.getInstance().removeActivity(this);
    }

}

File: app/src/main/java/com/dsy/mvp/base/BaseActivity.kt
package com.dsy.mvp.base

import android.app.Activity
import android.app.ProgressDialog
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import com.blankj.utilcode.util.ToastUtils
import com.dsy.mvp.base.impl.IPresenter
import com.dsy.mvp.base.impl.IView
import com.dsy.mvp.ui.adapter.NavIconType
import com.dsy.mvp.ui.adapter.ToolbarAdapter
import com.dsy.mvp.widget.status.LoadStatus
import com.dsy.mvp.widget.status.StatusHolder
import com.dylanc.loadinghelper.LoadingHelper
import com.dylanc.loadinghelper.ViewType
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

abstract class BaseActivity<T : IPresenter>(private val layoutId: Int) : AppCompatActivity(), IView {
    protected val TAG = this.javaClass.simpleName
    var progressDialog: ProgressDialog? = null
    protected lateinit var mPresenter: T
    protected lateinit var mLoadingHelper: LoadingHelper
    private lateinit var mStatusHolder: StatusHolder.Holder

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        beforeCreate()
        if (layoutId != 0) {
            setContentView(layoutId)
            initStatusHolder()
            mLoadingHelper = LoadingHelper(this)
            mStatusHolder.withRetry { onReload() }
        }
        mPresenter = initPresenter()
        mPresenter.attachView(this)
        initView()
        initEvent()
        initData()
    }

    private fun initStatusHolder() {
        mStatusHolder = if (getStatusWrapView() != null) {
            StatusHolder.getDefault().wrap(getStatusWrapView())
        } else {
            StatusHolder.getDefault().wrap(this)
        }
    }

    /**
     * 状态布局包裹的view
     */
    protected open fun getStatusWrapView(): View? {
        return null
    }
    override fun showLoadingView() {
        mStatusHolder.showLoading()
    }

    override fun showContentView() {
        mStatusHolder.showLoadSuccess()
    }

    override fun showErrorView() {
        mStatusHolder.showLoadFailed()
    }

    override fun showEmptyView() {
        mStatusHolder.showEmpty()
    }

    override fun showCustomView(status: LoadStatus) {
        mStatusHolder.showLoadingStatus(status)
    }


    protected fun setToolbar(title: String, type: NavIconType = NavIconType.NONE) {
        mLoadingHelper.apply {
            register(ViewType.TITLE, ToolbarAdapter(title, type))
            setDecorHeader(ViewType.TITLE)
        }
    }

    protected open fun onReload() {}

    @Suppress("UNCHECKED_CAST")
    protected fun initPresenter(): T {
        val type: Type? = javaClass.genericSuperclass
        val modelClass = if (type is ParameterizedType) {
            type.actualTypeArguments[0] as Class<*>
        } else {
            //如果没有指定泛型参数，则默认使用EmptyPresenter
            SimplePresenter::class.java
        }
        return modelClass.newInstance() as T
    }


    /**
     * 布局载入之前
     */
    protected open fun beforeCreate() {}

    /**
     * 控件绑定
     */
    protected abstract fun initView()

    /**
     * 事件触发绑定
     */
    protected abstract fun initEvent()

    /**
     * 数据初始化
     */
    protected abstract fun initData()

    override val mActivity: Activity
        get() = this

    override fun toast(@StringRes id: Int) {
        ToastUtils.showShort(id)
    }

    override fun toast(msg: CharSequence) {
        ToastUtils.showShort(msg)
    }

    override fun showLoadingDialog() {
        progressDialog = ProgressDialog(this)
        progressDialog?.setMessage("加载中")
        progressDialog?.show()
    }

    override fun hideLoadingDialog() {
        progressDialog?.dismiss()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.detachView()
    }
}